<a href="https://colab.research.google.com/github/fazamumtaz/kelasbagus3-machine-learning/blob/main/tugas1_kelasbagus3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Import Library

In [146]:
import pandas as pd
from sklearn.preprocessing import MinMaxScaler

# munculin semua columns
pd.set_option('display.max_columns', None)

import library dan set_option supaya semua column nya terlihat, gaada column yang ketutup karena columnya terlalu lebar

## Load Dataset

In [147]:
df = pd.read_csv('https://raw.githubusercontent.com/fazamumtaz/kelasbagus3-machine-learning/refs/heads/main/datasets/Week1_Travel.csv')

In [168]:
df.head(5)

additional_info  \
0  Room Service|Internet Access|Restaurant|Free I...   
1                               Room Service|Gym/Spa   
2                           Restaurant|Swimming Pool   
3                                                NaN   
4                         Internet Access|Restaurant   

                                             address            area     city  \
0  15th Mile, N.H.21,Manali, District Kullu,Himac...          Others   Manali   
1  A-585, Sushant Lok-1 ,Near Iffco Chowk Metro S...     Sushant Lok  Gurgaon   
2  Cobra Vaddo,Calungate Baga Road, Bardez, Calan...  Calangute Area      Goa   
3                                              Simsa   Village Simsa   Manali   
4         8180 Street No.-6,Arakashan Road,Paharganj       Paharganj    Delhi   

  country  crawl_date  guest_recommendation hotel_brand hotel_category  \
0   India  2016-07-24                  85.0         NaN        gostays   
1   India  2016-07-24                  87.0         NaN        regular   
2   India  2016-07-24                  50.0         NaN        regular   
3   India  2016-07-24                 100.0         NaN        regular   
4   India  2016-07-24                  63.0         NaN        regular   

                                   hotel_description  \
0  The standard check-in time is 12:00 PM and the...   
1  The standard check-in time is 12:00 PM and the...   
2  The standard check-in time is 12:00 PM and the...   
3  The standard check-in time is 12:00 PM and the...   
4  The standard check-in time is 12:00 PM and the...   

                                    hotel_facilities  hotel_star_rating  \
0  Doctor on Call|Dry Cleaning|Laundry Service Av...                  2   
1  Airport Transfer Available / Surcharge|Banquet...                  0   
2  Swimming Pool|Bar / Lounge |Laundry Service Av...                  0   
3  Doctor on Call|Dry Cleaning|Laundry Service Av...                  2   
4  Internet Access - Surcharge|Laundry Service Av...                  2   

   image_count   latitude        locality  longitude  \
0           20  32.139387          Others  77.154660   
1           17  28.472097     Sushant Lok  77.072546   
2           28  15.548398  Calangute Area  73.757634   
3            4  32.223603   Village Simsa  77.185900   
4            6  28.646777       Paharganj  77.212735   

                                             pageurl  \
0  http://www.goibibo.com/hotels/detail/Manali/hn...   
1  http://www.goibibo.com/hotels/detail/Gurgaon/h...   
2  http://www.goibibo.com/hotels/detail/Goa/hn/87...   
3  http://www.goibibo.com/hotels/detail/Manali/hn...   
4  http://www.goibibo.com/hotels/detail/Delhi/hn/...   

                                   point_of_interest       property_id  \
0  Hadimba Temple|Naggar Village|Himalayan Nyingm...  HTLGBO1000016238   
1  Sushant Lok|Sahara Mall|Amity International Sc...  HTLGBO1000015822   
2  Anjuna Beach|Calangute Beach|Titos lane, baga|...  HTLGBO1000004551   
3  The Mall|Tibetian Monastery|Vashisht Hot Baths...  HTLGBO1000007993   
4  Gaffar Market|YMCA Institute|Agrasen Boali|Cha...  HTLGBO1000008281   

         property_name property_type province                        qts  \
0     Baragarh Regency        Resort   Manali  2016-07-24 01:12:30 +0000   
1  Asian Suites A- 585   Guest House  Gurgaon  2016-07-24 01:12:30 +0000   
2        Bevvan Resort        Resort      Goa  2016-07-24 01:12:30 +0000   
3    Apple Inn Cottage       Cottage   Manali  2016-07-24 01:12:30 +0000   
4  Anmol Hotel Pvt.Ltd         Hotel    Delhi  2016-07-24 01:12:30 +0000   

            query_time_stamp  \
0  2016-07-24 01:12:30 +0000   
1  2016-07-24 01:12:30 +0000   
2  2016-07-24 01:12:30 +0000   
3  2016-07-24 01:12:30 +0000   
4  2016-07-24 01:12:30 +0000   

                            review_count_by_category       room_area  \
0  positive reviews::74|critical reviews::13|revi...             NaN   
1  positive reviews::8|critical reviews::0|review...  14 X 12 Sq.ft

load dataset, dan pake method `.head()` buat liat datanya sekilas

## Data Understanding

In [149]:
df.shape

(4000, 36)

In [150]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 36 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   additional_info           3192 non-null   object 
 1   address                   4000 non-null   object 
 2   area                      3965 non-null   object 
 3   city                      4000 non-null   object 
 4   country                   4000 non-null   object 
 5   crawl_date                4000 non-null   object 
 6   guest_recommendation      2416 non-null   float64
 7   hotel_brand               389 non-null    object 
 8   hotel_category            4000 non-null   object 
 9   hotel_description         3983 non-null   object 
 10  hotel_facilities          3806 non-null   object 
 11  hotel_star_rating         4000 non-null   int64  
 12  image_count               4000 non-null   int64  
 13  latitude                  4000 non-null   float64
 14  locality

## Data Preprocessing

### Hapus kolom yang sekiranya kurang relevan

In [167]:
pd.DataFrame({
    'Sum of MissVal': df[['hotel_brand', 'room_area']].isnull().sum(),
    'Percentage of MissVal': df[['hotel_brand', 'room_area']].isnull().sum() / len(df[['hotel_brand', 'room_area']]) * 100
})

Sum of MissVal  Percentage of MissVal
hotel_brand            3611                 90.275
room_area              2872                 71.800

In [151]:
df_dropped = df.drop(['sitename', 'qts', 'query_time_stamp',
                      'crawl_date', 'pageurl', 'similar_hotel',
                      'uniq_id', 'hotel_description', 'country',
                      'address', 'hotel_brand', 'room_area' ], axis=1)

Berikut alesan kenapa columns di atas itu di-drop:

1. `sitename`, `qts`, `query`, `crawl_date`, `pageurl`, `similar_hotel`, `uniq_id`: data-data ini kemungkinan besar adalah data hasil web scraping, jadi ada kolom kaya `crawl_date` dan yang lain. sementara kolom `pageurl`, `similar_hotel` itu menurutku kayanya ngga begitu relevan
2. `hotel_description`, `address`,: ini kaya free text yang kompleks banget, gaada pola nya, jadi kayanya juga useless kalo dikasih ke model.
3. `hotel_brand`, `room_aera`: banyak sekali missing values di data ini, `hotel_brand` dengan persentase **90.2%** missing values dan `room_area` dengan persentase missing values **71.8%**
4. `country`sedangkan untuk country, karean semuanya itu ada di India, jadi datanya ngga terlalu spesial

### Handling Missing Values

In [163]:
missing_values = pd.DataFrame({
    'Sum of MissVal': df_dropped.isnull().sum(),
    'Percentage of MissVal': df_dropped.isnull().sum() / len(df_dropped) * 100
})
missing_values

Sum of MissVal  Percentage of MissVal
additional_info                      808                 20.200
area                                  35                  0.875
city                                   0                  0.000
guest_recommendation                1584                 39.600
hotel_category                         0                  0.000
hotel_facilities                     194                  4.850
hotel_star_rating                      0                  0.000
image_count                            0                  0.000
latitude                               0                  0.000
locality                              35                  0.875
longitude                              0                  0.000
point_of_interest                    240                  6.000
property_id                            0                  0.000
property_name                          0                  0.000
property_type                          0                  0.000
province                               0                  0.000
review_count_by_category            1585                 39.625
room_count                             0                  0.000
room_facilities                      270                  6.750
room_type                              0                  0.000
site_review_count                   1584                 39.600
site_review_rating                  1584                 39.600
site_stay_review_rating                0                  0.000
state                                  0                  0.000

In [197]:
df_clean = df_dropped.copy()

cols_to_unknown = ['additional_info', 'area',
                   'locality', 'point_of_interest',
                  #  'review_count_by_category',
                   'room_facilities', 'hotel_facilities']
cols_to_zero = ['guest_recommendation', 'site_review_count', 'site_review_rating']

for col in cols_to_unknown:
    df_clean[col] = df_clean[col].fillna('Unknown')

for col in cols_to_zero:
  df_clean[col] = df_clean[col].fillna(0)

buat column `object` yang NaN, itu diubah jadi Unknown, dan column yang data type nya `int` atau `float`, diubah jadi 0. Semenjak Kak Bagus bilang lebih mending salah input nilai daripada Mengada-ada nilai yang sebenernya gaada, aku jadi takut pake Mean, Median, atau Mode gitu HEHEHE

In [193]:
df_clean.isnull().sum()

additional_info                0
area                           0
city                           0
guest_recommendation           0
hotel_category                 0
hotel_facilities               0
hotel_star_rating              0
image_count                    0
latitude                       0
locality                       0
longitude                      0
point_of_interest              0
property_id                    0
property_name                  0
property_type                  0
province                       0
review_count_by_category    1585
room_count                     0
room_facilities                0
room_type                      0
site_review_count              0
site_review_rating             0
site_stay_review_rating        0
state                          0
dtype: int64

cihuy bersih

In [194]:
df_clean.loc[1, 'review_count_by_category']

'positive reviews::8|critical reviews::0|reviews with images::2'

In [196]:
def fix_site_stay(srs, col):
  rating = [r.split("::")[-1] for r in srs[col].split("|")]
  rating = [float(r) if len(r) > 0 else 0 for r in rating]
  return rating

ratings = pd.DataFrame(data=df_clean.apply(lambda x: fix_site_stay(x, 'site_stay_review_rating'),
                                                           axis='columns').tolist(),
                       columns=['service_quality_rating', 'amenities_rating',
                                'food_and_drinks_rating', 'value_for_money_rating',
                                'location_rating', 'cleanliness_rating'])